### ISSUE TITLE
New compilers break on potentially passing NULL to printf

### ISSUE THREAD

**Issue by cpaelzer** [mdbtools/mdbtools#352]
Function `const char * mdb_get_objtype_string(int obj_type)` can be null in certain cases as it has an explicit return NULL if it seems to be out of range.

Newer compilers break here, and when checking the detail it is a new error:
```
In function ‘printf’,
    inlined from ‘mdb_dump_catalog’ at catalog.c:195:4:
/usr/include/powerpc64le-linux-gnu/bits/stdio2.h:112:10: error: ‘%-12s’ directive argument is null [-Werror=format-overflow=]
  112 |   return __printf_chk (__USE_FORTIFY_LEVEL - 1, __fmt, __va_arg_pack ());
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I don't yet know why, but so far I only see this issue on Ubuntu builds on ppc64
https://launchpadlibrarian.net/560405626/buildlog_ubuntu-impish-ppc64el.mdbtools_0.9.4-1_BUILDING.txt.gz
Maybe the sizes or the amount it can evaluate at compile time differs.

The problem is the potential return of NULL, for debugging I switched it to `char *)0x1` and it built fine.
So this isn't any runtime issue, but the theoretical chance that NULL might be returned and eventually end up as argument for printf `%-12s` which will fail at that time.

This is due to the pass of the result to printf unchecked:
```
mdb_dump_catalog(MdbHandle *mdb, int obj_type)
{
	unsigned int i;
	MdbCatalogEntry *entry;

	mdb_read_catalog(mdb, obj_type);
	for (i=0;i<mdb->num_catalog;i++) {
                entry = g_ptr_array_index(mdb->catalog,i);
		if (obj_type==MDB_ANY || entry->object_type==obj_type) {
			printf("Type: %-12s Name: %-48s Page: %06lx\n",
			mdb_get_objtype_string(entry->object_type),
```

You could either overhaul that generally, or as an alternative just check for NULL before using it.
An example (but feel free to use any alteration thereof that suits your coding style preferences better)
```
--- catalog.c.orig	2021-09-29 13:24:12.168071214 +0000
+++ catalog.c	2021-09-29 13:24:16.535907721 +0000
@@ -192,8 +192,11 @@
 	for (i=0;i<mdb->num_catalog;i++) {
                 entry = g_ptr_array_index(mdb->catalog,i);
 		if (obj_type==MDB_ANY || entry->object_type==obj_type) {
+			const char *name = mdb_get_objtype_string(entry->object_type);
+		        if (name == NULL)
+				name = "Unkown entry";
 			printf("Type: %-12s Name: %-48s Page: %06lx\n",
-			mdb_get_objtype_string(entry->object_type),
+			name,
 			entry->object_name,
 			entry->table_pg);
 		}
```

**Comment by evanmiller** [comment]
Thanks for the report. Stylistically the easiest solution will be the "Elvis operator":

```
printf("Type: %-12s Name: %-48s Page: %06lx\n",
mdb_get_objtype_string(entry->object_type) ?: "Unknown"
```

This will give more consistent behavior than whatever is happening now with NULL pointers.

If there's a compiler flag to turn on this check, then we can address it systematically.

**Comment by cpaelzer** [comment]
Hi @evanmiller - thanks for the feedback on preferences.
I have used that, test built and opened two PRs #353 and #354

The flag to turn it on is in https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html see `format-overflow`. AFAICS you'll also need GCC-11 to trigger this particular issue that I've reported. And as I mentioned for no reasons I'd see yet it only gets to that detection on ppc64, maybe there the involves sizeof behaves slightly different?
